src/download.rs

```diff
@@ -549,3 +572,176 @@ mod error {
+        let path = self.as_ref();
+        let file = OpenOptions::new()
+            .read(true)
+            .open(path)
+            .await
+            .context(error::OpenFile { path })?;
+
+        let mut block_device_size = 0;
+        let result = unsafe { ioctl::blkgetsize64(file.as_raw_fd(), &mut block_device_size) }
+            .context(error::GetBlockDeviceSize)?;
+        ensure!(result == 0, error::InvalidBlockDeviceSize { result });
+
+        let block_device_size =
+            i64::try_from(block_device_size).with_context(|| error::ConvertNumber {
+                what: "block device size",
+                number: block_device_size.to_string(),
+                target: "i64",
+            })?;
+
+        // Make sure the block device is big enough to hold the snapshot
+        ensure!(
+            block_device_size >= length,
+            error::BlockDeviceTooSmall {
+                block_device_size: block_device_size / GIBIBYTE,
+                needed: length / GIBIBYTE,
+            }
+
+        );
+
+        Ok(())
+    }
+
+    // returns the file path to which blocks must be written
+    fn write_path(&self) -> Result<&Path> {
+        Ok(self.as_ref())
+    }
+
+    // no-op
+    fn finalize(&mut self) -> Result<()> {
+        Ok(())
+    }
+}
+
+struct FileTarget {
+    path: PathBuf,
+    temp_file: Option<NamedTempFile>,
+}
+
+impl FileTarget {
+    async fn new<P: AsRef<Path>>(path: P) -> Result<Box<dyn SnapshotWriteTarget>> {
+        let path = path.as_ref();
+        Ok(Box::new(FileTarget {
+                path: path.into(),
+                temp_file: None,
+                }
+            ))
+    }
+}
+
+impl AsRef<Path> for FileTarget {
+    fn as_ref(&self) -> &Path {
+        self.path.as_ref()
+    }
+}
+
+#[async_trait]
+impl SnapshotWriteTarget for FileTarget {
+    // truncate file to desired size
+    async fn grow(&mut self, length: i64) -> Result<()> {
+        let path = self.as_ref();
+
+        // Create a temporary file and extend it to the required size.
+        let target_dir = path
+            .parent()
+            .context(error::ValidateParentDirectory { path })?;
+
+        let temp_file = NamedTempFile::new_in(target_dir)
+            .context(error::CreateTempFile { path: target_dir })?;
+
+        let temp_file_len = length;
+        let temp_file_len = u64::try_from(temp_file_len).with_context(|| error::ConvertNumber {
+            what: "temp file length",
+            number: temp_file_len.to_string(),
+            target: "u64",
+        })?;
+
+        temp_file
+            .as_file()
+            .set_len(temp_file_len)
+            .context(error::ExtendTempFile {
+                path: temp_file.as_ref(),
+                }
+            )?;
+
+        self.temp_file.replace(temp_file);
+
+        Ok(())
+    }
+
+    fn write_path(&self) -> Result<&Path> {
+        Ok(self.temp_file.as_ref().unwrap().as_ref())
+    }
```

Comment:
We should avoid `unwrap()` which will cause a panic if we have a `None`.

```suggestion
    fn write_path(&self) -> Result<&Path> {
        let path = self.temp_file.as_ref().context(error::MissingTempFile{})?;
        Ok(path.as_ref())
    }
```

Reply:
Thanks. I have made the change.